fix(docs): add release notes v4 (#3139)
patternfly/patternfly: UPGRADE-GUIDE.md

Various colors have been updated to increase contrast in more usage scenarios to meet accessibility requirements. Some color variables have been remapped to better suit new component designs such as alerts, and we’ve also added new `-50` color tints to serve as colored backgrounds where appropriate. As a followup to last year's color changes, we've updated Blue and Cyan palettes' `-600` and `-700` values as well.

### Mobile first CSS
We've updated some components' CSS to be mobile first by using `@min-width` media queries. The components that are now mobile-first are: background image, data list, toolbar, form, login, page, toolbar, and wizard [(#2816)](https://github.com/patternfly/patternfly/pull/2816).
  * All components are now mobile-first, but one component that cannot satisfy this is the table, specifically with use of the `.pf-m-grid[-on-{breakpoint}]` modifers. At the specified grid breakpoint, native/default properties are modified to achieve a grid based layout. To overwrite these changes would require adding back native properties to the table. For that reason, the table will remain the exception to a “mobile-first” approach.

### Vertical navigation hidden breakpoint

> We've changed the breakpoint at which we hide the vertical nav to be `$pf-global--breakpoint--xl` (1200px) rather than `$pf-global--breakpoint--md` (768px) [(#2962)](https://github.com/patternfly/patternfly/pull/2962). The overall page chrome and some individual components' horizontal spacing becomes more compact at this breakpoint (true of the old breakpoint and the new), so applications may find that custom elements that aligned with the old breakpoint may not align now since the chrome spacing changes at 1200px instead of 768px. An application may need to make updates to match the chrome's spacing at the new breakpoint.

Reviewer: ```suggestion
We've changed the hidden breakpoint for the vertical nav to be `$pf-global--breakpoint--xl` (1200px) rather than `$pf-global--breakpoint--md` (768px) [(#2962)](https://github.com/patternfly/patternfly/pull/2962). The overall page chrome and some individual components' horizontal spacing becomes more compact at this breakpoint (true of the old breakpoint and the new). Applications' custom elements that aligned with the old breakpoint may not align now since the chrome spacing changes at 1200px instead of 768px. You may need to make updates to match the chrome's spacing at the new breakpoint.
```

### Default font
We've updated the default font to be `RedHatText` instead of `Overpass` [(#2955)](https://github.com/patternfly/patternfly/pull/2955). To use the old font, add the class `pf-m-overpass-font` on an element that wraps your application (ie, `<body>`).

PatternFly has chosen to adopt a single font (`RedHatText`) across the library [(#3023)](https://github.com/patternfly/patternfly/pull/3023). Previously, when having opted in to the Red Hat font via `.pf-m-redhat-font`, PatternFly used 2 fonts - `RedHatText` (RHT) defined as `--pf-global--FontFamily--sans-serif`, and `RedHatDisplay` (RHD) defined as `--pf-global--FontFamily--heading--sans-serif`. `--pf-global--FontFamily--heading--sans-serif` has been removed and components have been updated to use `--pf-global--FontFamily--sans-serif` instead.

> Users do not have to do anything further to consume this change. However, with the change from `Overpass` to `RedHatText`, and with the removal of `RedHatDisplay`, we encourage teams to review their application's typography styles to ensure they are correct.

Reviewer 2: LGTM!

- Added new cyan, gold, green, red colors [(#2906)](https://github.com/patternfly/patternfly/pull/2906)

Removed classes:
- `.pf-m-redhat-font`
  - `RedHatText` is now the default font. Apply `.pf-m-overpass-font` to use `Overpass` instead.

Removed variables:
- `--pf-global--BackgroundColor--150` and `--pf-global--BackgroundColor--300`
  - These can be updated to use `--pf-global--BackgroundColor--200` instead. However you might want to consult your visual designer first.
- All bootstrap variables [(#2871)](https://github.com/patternfly/patternfly/pull/2871)
  - These are no longer included with PatternFly by default. If you still need these, you can import `sass-utilities/bs-variables.scss` manually.
- `--pf-global--link--FontWeight`
- `--pf-global--FontFamily--redhatfont--sans-serif`
- `--pf-global--FontFamily--redhatfont--heading--sans-serif`
- `--pf-global--FontFamily--redhatfont--monospace`
- `--pf-global--FontWeight--redhatfont--bold`
- `--pf-global--golden-ratio`

## Components

**CSS variables:**
Our components have seen many CSS variables changed, removed, and renamed. Removed and renamed variables may require changes to your CSS and HTML.

**Visual changes:**

> Many of our components have had design updates that introduce visual breaking changes. Some are simple updates to existing styles (e.g., spacing or color), while others are an overhaul of the design (e.g., Tabs and Label).

Author: I like TitleCase, so I'll leave them be!

- `--pf-c-alert--m-inline__icon--PaddingTop`
- `--pf-c-alert--m-inline__icon--PaddingRight`
- `--pf-c-alert--m-inline__icon--PaddingBottom`
- `--pf-c-alert--m-inline__icon--PaddingLeft`
- `--pf-c-alert--m-inline--m-warning__icon--FontSize`
- `--pf-c-alert--m-inline--m-success__icon--Color`
- `--pf-c-alert--m-inline--m-success--before--BackgroundColor`
- `--pf-c-alert--m-inline--m-danger__icon--Color`
- `--pf-c-alert--m-inline--m-danger--before--BackgroundColor`
- `--pf-c-alert--m-inline--m-warning__icon--Color`
- `--pf-c-alert--m-inline--m-warning--before--BackgroundColor`
- `--pf-c-alert--m-inline--m-info__icon--Color`
- `--pf-c-alert--m-inline--m-info--before--BackgroundColor`

### Application launcher
- Applied external link icon class to icon wrapper [(#2904)](https://github.com/patternfly/patternfly/pull/2904)
- Removed separator in favor of divider component [(#2944)](https://github.com/patternfly/patternfly/pull/2944)
- Removed hover, active, focus, and disabled from workspace [(#2975)](https://github.com/patternfly/patternfly/pull/2975)
- Cleaned up variables [(#3012)](https://github.com/patternfly/patternfly/pull/3012)

Updated markup:

> - Moves the class `.pf-c-app-launcher__menu-item-external-icon` from the external link icon itself to a `<span>` that wraps the external link icon. Any instances of the external link icon should be wrapped in a `<span>` and the `.pf-c-app-launcher__menu-item-external-icon` class should be moved from the icon to the `<span>`.

Reviewer 2: yep 👍

- `.pf-m-hover` from `.pf-c-app-launcher__toggle` and `.pf-c-app-launcher__menu`
- `.pf-m-focus` from `.pf-c-app-launcher__toggle` and `.pf-c-app-launcher__menu`
- `.pf-m-disabled` from `.pf-c-app-launcher__toggle`
  - The `:hover`, `:active`, `:focus`, and `:disabled` selectors still have styles applied to them

Renamed variables:
- `--pf-c-app-launcher--m-top__menu--Transform` to `--pf-c-app-launcher--m-top__menu--TranslateY`
- `--pf-c-app-launcher__menu-item-external-icon--Transform` to `--pf-c-app-launcher__menu-item-external-icon--TranslateY`

Removed variables:
- `--pf-c-app-launcher__group--PaddingTop`
- `--pf-c-app-launcher__group--first-child--PaddingTop`
- `--pf-c-app-launcher__menu--BorderWidth`
- `--pf-c-app-launcher__separator--Height`
- `--pf-c-app-launcher__separator--BackgroundColor`
- `--pf-c-app-launcher__separator--MarginTop`
- `--pf-c-app-launcher__separator--MarginBottom`
- `--pf-c-app-launcher__separator--Height`

### Backdrop
- Removed blur [(#3009)](https://github.com/patternfly/patternfly/pull/3009)

Renamed variables:
- `--pf-c-backdrop--Color` to `--pf-c-backdrop--BackgroundColor`

Removed variables:
- `--pf-c-backdrop--BackdropFilter`

### Breadcrumb

> - Added word-break, display link and icon inline, move divider [(#2916)](https://github.com/patternfly/patternfly/pull/2916)

Reviewer 2: It's just the [PR title/commit message](https://github.com/patternfly/patternfly/pull/2916). What it actually does is:

add word-break - enables long strings as breadcrumb items to break to multiple lines if needed (like a really long word)

display link and icon inline - displays the divider icon and breadcrumb item text inline with (beside) one another when breadcrumb items wrap to multiple lines as opposed to allowing the divider and breadcrumb item text to wrap independently of one another.

move divider - dividers were moved to come before the breadcrumb item link instead of after.

Also, the notes from the PR about these 3 things:

> This PR changes the visual appearance of breadcrumbs when they break to multiple lines.
> * The text will now break if there is a long string in an item link that is wider than the viewport
> * The divider is displayed before the link text, instead of after it
> * The divider will also always appear inline with the item text instead of wrapping to a new line in some cases.

> - Added word-break, display link and icon inline, move divider [(#2916)](https://github.com/patternfly/patternfly/pull/2916)

Author: Updated to use those 3 bullet points!

- Removed hover, active, focus, and disabled from workspace [(#2975)](https://github.com/patternfly/patternfly/pull/2975)

> - Made redhat font default, add opt-in for overpass [(#2955)](https://github.com/patternfly/patternfly/pull/2955)

Author: Line 244 is copy/pasted 15 times for various components. I removed line 245 (which is redundant) and reworded the 15 occurrences and line 243.

> - Made redhat font default, add opt-in for overpass [(#2955)](https://github.com/patternfly/patternfly/pull/2955)

Reviewer 2: For the lines like this at the beginning of notes about a component, that text is the commit message of the PR. As long as we're OK with that text not matching the PR commit message, what about:

> - Enabled long strings to break to multiple lines, made item link and divider icon display inline, moved divider icon to come before the item link text
> - Removed hover, active, focus, and disabled from workspace 
> -  Made redhat font default, added opt-in for overpass

Updated markup:

> - Any reference to the divider element `.pf-c-breadcrumb__item-divider` will need to be moved in the markup to come before the breadcrumb link element `.pf-c-breadcrumb__link` instead of after it.

Reviewer: ```suggestion
- In the markup, move any reference to the divider element `.pf-c-breadcrumb__item-divider` so that is comes before the breadcrumb link element `.pf-c-breadcrumb__link` instead of after it.
```

Updated CSS:
- The text will now break if there is a long string in an item link that is wider than the viewport.
- The divider will always appear inline with the item text instead of wrapping to a new line in some cases.

Removed classes:
- `.pf-m-hover` from `.pf-c-breadcrumb__link`
  - The `:hover` selector still has styles applied to it.

Renamed variables:
- `--pf-c-breadcrumb__item-divider--MarginLeft` to `--pf-c-breadcrumb__item-divider--MarginRight`

Removed variables:
- `--pf-c-breadcrumb__item--FontWeight`
- `--pf-c-breadcrumb__link--FontWeight`

### Button
- Removed hover, active, focus, and disabled from workspace [(#2975)](https://github.com/patternfly/patternfly/pull/2975)
- Cleaned up vars [(#3028)](https://github.com/patternfly/patternfly/pull/3028)
- Updated control variation disabled state [(#3049)](https://github.com/patternfly/patternfly/pull/3049)
- Added modifiers to position icon in button [(#2828)](https://github.com/patternfly/patternfly/pull/2828)
- Made shield styles optional by default [(#2872)](https://github.com/patternfly/patternfly/pull/2872)

Updated markup:

> - Adds the modifiers `.pf-m-start` and `.pf-m-end` to position `pf-c-button__icon` in the button component depending on whether it comes before or after text. If you are using a button with an icon, you will need to use one of these modifiers to style the icon.

Reviewer: ```suggestion
- Adds the modifiers `.pf-m-start` and `.pf-m-end` to position `pf-c-button__icon` in the button component depending on whether it comes before or after text. If you are using a button with an icon, use one of these modifiers to style the icon.
``` 
Is "Adds" supposed to be "Added" ?

Author: Yes!

Reviewer 2: yep. A lot of these blurbs probably came from PR description notes written like "This PR adds the modifiers..." But for consistency in tense in the release notes, they should be updated.

Updated CSS:
- Link and plain buttons now define `background-color: transparent` for their normal, hover, focus, and active states.
- The control button now sets a white background and default text color for its normal, hover, focus, and active states.

> - The `:active` CSS selectors in the button component's stylesheet were moved after `:focus`, so that if the `:focus` and `:active` pseudo selectors declare styles on the same CSS properties for an element, the `:active` styles will now be visible when the element is `:active`. No changes are needed to consume this change, however any `:active` styles that may have been overridden by the `:focus` styles coming later in the stylesheet may now be visible.

Reviewer: ```suggestion
- The `:active` CSS selectors in the button component's stylesheet were moved after `:focus` so that if the `:focus` and `:active` pseudo selectors declare styles on the same CSS properties for an element, the `:active` styles will now be visible when the element is `:active`. No changes are needed to consume this change, but any `:active` styles that may have been overridden by the `:focus` styles coming later in the stylesheet may now be visible.
``` 
"No changes are necessary to consume this change" sounds awkward. Can we reword to something that's more specific and conversational?

Author: This case seems rather far-fetched and not useful for an upgrade guide. I've removed it.

Reviewer 2: You used "You don't have to do anything further" above, which I think works well. 

What about

> You don't have to do anything further to get this update, but any `:active`...

?

Removed classes:
- `.pf-c-button__text` - element was removed
- `.pf-m-hover` from `.pf-c-button`
- `.pf-m-focus` from `.pf-c-button`

>   - The `:hover` and `:focus` selectors still have styles applied to them.

Reviewer: I recommend making 282 and 283 consistent in structure with the other line elements. It isn't clear what they mean, especially since we say "Removed classes." Maybe we mean "Changed `.pf-m-hover` to `.pf-c-button` " or "Removed `.pf-m-hover` from `.pf-c-button` " ?

Reviewer 2: Those lines are meant to communicate "Removed .pf-m-hover/pf-m-focus from .pf-c-button"

- `.pf-c-card__head-main` has been renamed to `.pf-c-card__header-main`.

Renamed variables:
- `--pf-c-card--m-compact__header--not-last-child--PaddingBottom` to `--pf-c-card--m-compact__header--not--last-child--PaddingBottom`
- `--pf-c-card__header--not-last-child--PaddingBottom` to `--pf-c-card__header--not--last-child--PaddingBottom`
- `--pf-c-card__header--not-last-child--PaddingBottom` to `--pf-c-card__header--not--last-child--PaddingBottom`
- `--pf-c-card--m-compact__header--not--last-child--PaddingBottom` to `--pf-c-card--m-compact__title--not--last-child--PaddingBottom`
- `--pf-c-card__header--FontFamily` to `--pf-c-card__title--FontFamily`
- `--pf-c-card__header--FontWeight` to `--pf-c-card__title--FontWeight`
- `--pf-c-card__header--not--last-child--PaddingBottom` to `--pf-c-card__title--not--last-child--PaddingBottom`
- `--pf-c-card__header--not--last-child--PaddingBottom` to `--pf-c-card__title--not--last-child--PaddingBottom`

### Chip
- Refactor styles [(#2941)](https://github.com/patternfly/patternfly/pull/2941).
- Removed hover, active, focus, and disabled from workspace [(#2975)](https://github.com/patternfly/patternfly/pull/2975).

> - Adjusted chip top/bottom padding, chip group margin in select [(#3079)](https://github.com/patternfly/patternfly/pull/3079).

Reviewer: ```suggestion
- Adjusted chip top/bottom padding and the chip group margin in select [(#3079)](https://github.com/patternfly/patternfly/pull/3079).
``` 
Did we adjust the chip top/bottom padding and the chip group margin in select? If so, edits applied are meant to add that clarity.

Reviewer 2: yep!

- `--pf-c-chip--m-overflow--c-button--focus--BorderColor` to `--pf-c-chip--m-overflow--focus--before--BorderColor`
- `--pf-c-chip--c-badge--MarginLeft` to `--pf-c-chip__c-badge--MarginLeft`

Removed variables:
- `--pf-c-chip--m-read-only--PaddingTop`
- `--pf-c-chip--m-read-only--PaddingRight`
- `--pf-c-chip--m-read-only--PaddingBottom`
- `--pf-c-chip--m-overflow--BackgroundColor`
- `--pf-c-chip--m-overflow--PaddingLeft`
- `--pf-c-chip--m-overflow--BorderWidth`
- `--pf-c-chip--m-overflow--c-button--BorderRadius`
- `--pf-c-chip--m-overflow--c-button--BorderWidth`
- `--pf-c-chip--m-overflow--c-button--PaddingLeft`
- `--pf-c-chip--m-overflow--c-button--PaddingRight`
- `--pf-c-chip--m-overflow--c-button--hover--BorderWidth`
- `--pf-c-chip--m-overflow--c-button--hover--BorderColor`
- `--pf-c-chip--m-overflow--c-button--active--BorderWidth`
- `--pf-c-chip--m-overflow--c-button--active--BorderColor`
- `--pf-c-chip--m-overflow--c-button--focus--BorderWidth`
- `--pf-c-chip--m-overflow--c-button--focus--BorderColor`
- `--pf-c-select__toggle-wrapper--c-chip--c-button--PaddingTop`
- `--pf-c-select__toggle-wrapper--c-chip--c-button--PaddingBottom`

### Chip group

> - Clarify chip group behavior [(#2961)](https://github.com/patternfly/patternfly/pull/2961)

Reviewer: Should "Clarify" be "Clarified" ?

Author: I just tracked the issue back to the design docs... I've rephrased to "Clarified chip group types, overflow, and categories"

- No longer nests [(#3058)](https://github.com/patternfly/patternfly/pull/3058)

Updated markup:
- The outer `.pf-c-chip-group` element should now always be a `<div>`. All instances of `.pf-c-chip-group` should be updated to be a `<div>`.
- `.pf-c-chip-group__label` is now a `<span>` instead of a heading element.
- `.pf-c-chip-group__label` should have an `id` value that will be used as the value for `aria-labelledby` on the `ul.pf-c-chip-group__list` element.

> - Since the text in `.pf-c-chip-group__label` is read to screen readers via `aria-labelledby` on the `pf-c-chip-group__list` element, `.pf-c-chip-group__label` should now also have `aria-hidden="true"`.

Reviewer: ```suggestion
- Because screen readers read the text in `.pf-c-chip-group__label` via `aria-labelledby` on the `pf-c-chip-group__list` element, `.pf-c-chip-group__label` should now also have `aria-hidden="true"`.
```

Author: My English teacher said never to start sentences with "Because."

I did it anyway.
